File: backend_python/chaos.py
killed_nodes = set()
killed_links = set()

def kill_node(node_id):
    killed_nodes.add(node_id)

def kill_link(node_a, node_b):
    key = "-".join(sorted([node_a, node_b]))
    killed_links.add(key)

def restore():
    killed_nodes.clear()
    killed_links.clear()

def get_state():
    return {
        'killedNodes': list(killed_nodes),
        'killedLinks': list(killed_links)
    }

File: backend/chaos.js
// chaos.js — Chaos engineering state: kill / restore nodes and links

const killedNodes = new Set();
const killedLinks = new Set();

/**
 * Build a canonical key for a link so that (A,B) and (B,A) are identical.
 */
function linkKey(a, b) {
  return [a, b].sort().join('-');
}

function killNode(nodeId) {
  killedNodes.add(nodeId);
}

function killLink(nodeA, nodeB) {
  killedLinks.add(linkKey(nodeA, nodeB));
}

function restore() {
  killedNodes.clear();
  killedLinks.clear();
}

function isNodeAlive(nodeId) {
  return !killedNodes.has(nodeId);
}

function isLinkAlive(nodeA, nodeB) {
  return !killedLinks.has(linkKey(nodeA, nodeB));
}

function getState() {
  return {
    killedNodes: Array.from(killedNodes),
    killedLinks: Array.from(killedLinks),
  };
}

module.exports = { killNode, killLink, restore, isNodeAlive, isLinkAlive, getState };
